server/dataone_register.py

```diff
@@ -263,47 +285,93 @@ def get_documents(package_pid):
 
 
-def get_package_title(docs):
-    for doc in docs:
-        if doc['formatType'] == 'METADATA':
-            return doc['title']
-    return ""
+def check_multiple_metadata(metadata):
+    if len(metadata) > 1:
+        raise RestException("Multiple documenting metadata objects found. "
+                            "This is unexpected and unhandled.")
 
 
-def get_package_list(path):
-    """
-    Retrieves a list of all the files in a package with
-    their sizes
-    """
+def get_package_list(path, package=None, isChild=False):
+    """"""
+    if package is None:
+        package = {}
 
     package_pid = get_package_pid(path)
     logger.debug('Found package PID of {}.'.format(package_pid))
 
     docs = get_documents(package_pid)
 
-    packageName = get_package_title(docs)
-
-    files = list()
-    for doc in docs:
-
-        dataMap = {
-            'name': get_object_name(doc),
-            'size': doc.get('size', -1),
-            'id': doc.get('identifier', -1),
-            'packageParent': packageName
-            }
-        files.append(dataMap)
-
-    return files
+    # Filter the Solr result by TYPE so we can construct the package
+    metadata = extract_metadata_docs(docs)
+    data = extract_data_docs(docs)
+    children = extract_resource_docs(docs)
+
+    # Verify what's in Solr is matching.
+    verify_results(package_pid, docs)
+
+    # Find the primary/documenting metadata so we can later on find the
+    # folder name
+    # TODO: Grabs the resmap a second time, fix this
+    documenting = get_documenting_identifiers(package_pid)
+
+    # Stop now if multiple objects document others
+    check_multiple_maps(documenting)
+
+    # Determine the folder name. This is usually the title of the metadata file
+    # in the package but when there are multiple metadata files in the package,
+    # we need to figure out which one is the 'main' or 'documenting' one.
+    primary_metadata = [doc for doc in metadata if 'documents' in doc]
+
+    check_multiple_metadata(primary_metadata)
+
+    data += [doc for doc in metadata if doc['identifier'] != primary_metadata[0]['identifier']]
+
+    fileList = get_package_files(data, metadata, primary_metadata)
+
+    # Add a new entry in the package structure
+    # if isChild:
+    #    package[-1][primary_metadata[0]['title']] = {'fileList': []}
+    # else:
+    package[primary_metadata[0]['title']] = {'fileList': []}
+
+    package[primary_metadata[0]['title']]['fileList'].append(fileList)
+    if children is not None and len(children) > 0:
+        for child in children:
+            get_package_list(child['identifier'], package[primary_metadata[0]['title']], True)
+    return package
+
+
+def get_package_files(data, metadata, primary_metadata):
+    fileList = {}
+    for fileObj in data:
+        try:
+            fileName = fileObj['fileName']
+        except KeyError:
+            fileName = fileObj['identifier']
+        try:
+            fileSize = fileObj['size']
```

Comment:
Same here: `fileName = int(fileObj.get('size', 0))`. Possibly catch `ValueError` if 'size' is non integer.